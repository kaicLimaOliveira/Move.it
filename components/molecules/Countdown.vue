<template>
  <div class="flex justify-center items-center mt-8 lg:mt-14 text-9xl text-title font-rajdhani">
    <CountdownDigits :digits="minutes" />
    <span class="bg-background px-2">:</span>
    <CountdownDigits :digits="seconds" />
  </div>
</template>

<script setup lang="ts">
import CountdownDigits from "~/components/atoms/CountdownDigits.vue";


const emit = defineEmits<{
  "completed": []
}>()

const { setTime, resetTime } = useCountdown();
const { time, isActive, minutes, seconds } = storeToRefs(useCountdown());


let TIMEOUT_REFERENCE: ReturnType<typeof setTimeout>;
function runCountdown(flag: boolean) {

  if (isActive && flag) {
    TIMEOUT_REFERENCE = setTimeout(() => setTime(time.value - 1), 1000);
  
  } else {
    clearTimeout(TIMEOUT_REFERENCE);
  }
}


watch(
  () => isActive.value,
  (newValue) => {
    runCountdown(newValue);

    if (!newValue) {
      resetTime();
    }
  }
)

watch(
  () => time.value,
  (newValue) => {
    if (newValue > 0) {runCountdown(true);}
    else {emit("completed")}
  }
)
</script>