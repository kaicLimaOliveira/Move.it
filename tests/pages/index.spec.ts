import { beforeEach, describe, expect, it, vi } from "vitest";
import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime'

import Countdown from "~/components/molecules/Countdown.vue";
import Index from "~/pages/index.vue";

vi.mock('~/utils', () => ({
  scrollToElement: vi.fn(),
  getRandomNumber: vi.fn(),
  playAudio: vi.fn(),
  sendNotification: vi.fn(),
  splitValue: vi.fn(),
}))

import { 
  getRandomNumber, 
  playAudio, 
  sendNotification,
} from '~/utils';

describe('page index', () => {
  const countdown = useCountdown();

  beforeEach(() => {
    global.Notification = {
			requestPermission: vi.fn(),
			permission: 'default',
		} as unknown as typeof Notification;
    
    useCountdown().$reset();
  });

  describe('Snapshots', () => {
    it('should render all child components and start a cycle button', async () => {
      const wrapper = await mountSuspended(Index);
      expect(wrapper.html()).toMatchSnapshot();
    })

    it('should render all child components and abandon cycle button', async () => {
			countdown.isActive = true;
			const wrapper = await mountSuspended(Index);
			expect(wrapper.html()).toMatchSnapshot();
		});

    it('should render all child components and abandon cycle button', async () => {
			countdown.hasCompleted = true;
			const wrapper = await mountSuspended(Index);
			expect(wrapper.html()).toMatchSnapshot();
		});
  })

  describe('Meta info', () => {
    it('should have a meta title', async () => {
      const wrapper = await mountSuspended(Index)
      // expect(wrapper.element.title).toBe('Home | move.it')
    })
  })

  describe('Mounted', () => {
    it('should request Notification permissions when mounted', async () => {
      await mountSuspended(Index);
      expect(global.Notification.requestPermission).toHaveBeenCalled();
    })

    it('should not request permissions if Notification doesnt exist', async () => {
      delete (global as any).Notification;
      await mountSuspended(Index);
      expect(global.Notification).toBe(undefined);
    })
  })
  
  describe('Button clicks', () => {
    it('should request Notification permissions when mounted', async() => {
      const wrapper = await mountSuspended(Index);
      const button = wrapper.find('button');
      await button.trigger('click');

      const newButtonRendered = wrapper.find('button');
      expect(newButtonRendered.text()).toBe('Abandonar ciclo');
    })

    it('should abandon cycle when abandon cycle button is clicked', async () => {
			const wrapper = await mountSuspended(Index);

			const button = wrapper.find('button');
			await button.trigger('click');

      const secondButton = wrapper.find('button');
			await secondButton.trigger('click');

			const newButton = wrapper.find('button');
			expect(newButton.text()).toBe('Iniciar ciclo');
		});
  })

  describe('Emitted events', () => {
		it('should run getNewChallenge and play audio and send notification', async () => {
			global.Notification = {
				requestPermission: vi.fn(),
				permission: 'granted',
			} as unknown as typeof Notification;

			mockNuxtImport('getRandomNumber', () => getRandomNumber);
      getRandomNumber(1, 1);

			const wrapper = await mountSuspended(Index);

			const setCurrentChallengeIndex = vi.spyOn(wrapper.vm, 'setCurrentChallengeIndex');

			const countdown = wrapper.findComponent(Countdown);
			await countdown.vm.$emit('completed');

			const button = wrapper.find('button');

			expect(button.text()).toBe('Ciclo completo');
			// expect(setCurrentChallengeIndex.mock).toBe(1);
			expect(playAudio).toHaveBeenCalledWith('./notification.mp3');
			expect(sendNotification).toHaveBeenCalledWith('New Challenge!', {
				body: 'A new challenge has started! Go complete it!',
				icon: './favicon.png',
			});
			// expect(scrollToElement).toBe('#challenge');
		});

		// it('should run getNewChallenge but doesnt play audio and doesnt send notification', async () => {
			
    //   const { getRandomNumber } = vi.hoisted(() => {
    //     return {
    //       getRandomNumber: vi.fn().mockImplementation(() => 1)
    //     }
    //   })

    //   mockNuxtImport('getRandomNumber', () => {
    //     return getRandomNumber
    //   })
      
    //   getRandomNumber.mockImplementation(() => 1)

		// 	const wrapper = await mountSuspended(Index);

		// 	const setCurrentChallengeIndex = vi.spyOn(wrapper.vm, 'setCurrentChallengeIndex');

		// 	const countdown = wrapper.findComponent(Countdown);
		// 	await countdown.vm.emit('completed');

		// 	const button = wrapper.find('button');

		// 	expect(button.text()).toBe('Ciclo completo');
		// 	expect(setCurrentChallengeIndex).toHaveBeenCalledWith(1);
		// 	expect(playAudio).not.toHaveBeenCalled();
		// 	expect(sendNotification).not.toHaveBeenCalled();
		// 	// expect(scrollToElement).toHaveBeenCalledWith('#challenge');
		// });
	});
})
