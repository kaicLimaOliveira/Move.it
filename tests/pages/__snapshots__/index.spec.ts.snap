// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page index > Snapshots > should render all child components and abandon cycle button 1`] = `
"<section class="flex flex-col lg:flex-row flex-1 lg:flex-none lg:mt-16 sm:gap-x-10 md:gap-x-20">
  <div class="flex flex-col w-full lg:w-1/2">
    <div class="flex items-center"><img src="https://github.com/kaicLimaOliveira.png" alt="Kaic Dev" class="mr-6 w-20 h-20 rounded-full">
      <div class="flex flex-col w-full"><strong class="text-title text-2xl">Kaic Dev</strong>
        <p class="flex items-center"><img src="/icons/level.svg" alt="level" class="mr-2"><span class="text-base">Level 1</span></p>
      </div>
    </div>
    <div class="flex justify-between items-center pb-5 border-b-2 border-gray-line mt-8 lg:mt-14 font-medium">
      <p class="text-lg"> Desafios Completos </p>
      <p class="text-2xl">0</p>
    </div>
    <div class="flex justify-center items-center mt-8 lg:mt-14 text-9xl text-title font-rajdhani">
      <p class="w-1/2 flex gap-x-1"></p><span class="bg-background px-2">:</span>
      <p class="w-1/2 flex gap-x-1"></p>
    </div><button class="button abandon"> Abandonar ciclo </button>
  </div>
  <div class="bg-white rounded-md flex flex-1 w-full mt-8 lg:mt-0 shadow-sm w-full lg:w-1/2" id="challenge">
    <section class="flex flex-col flex-1 justify-center items-center p-10 md:p-16">
      <h1 class="font-medium text-center text-2xl"> Finalize um ciclo para receber um desafio </h1><img src="/icons/level-up.svg" alt="Level up" class="mt-10 md:mt-16 min-h-[70px]">
      <p class="text-center mt-4"> Avance de Level completando desafios </p>
    </section>
    <transition-stub data-v-68482f27="" name="slide-fade" appear="false" persisted="false" css="true">
      <!--v-if-->
    </transition-stub>
  </div>
</section>"
`;

exports[`page index > Snapshots > should render all child components and abandon cycle button 2`] = `
"<section class="flex flex-col lg:flex-row flex-1 lg:flex-none lg:mt-16 sm:gap-x-10 md:gap-x-20">
  <div class="flex flex-col w-full lg:w-1/2">
    <div class="flex items-center"><img src="https://github.com/kaicLimaOliveira.png" alt="Kaic Dev" class="mr-6 w-20 h-20 rounded-full">
      <div class="flex flex-col w-full"><strong class="text-title text-2xl">Kaic Dev</strong>
        <p class="flex items-center"><img src="/icons/level.svg" alt="level" class="mr-2"><span class="text-base">Level 1</span></p>
      </div>
    </div>
    <div class="flex justify-between items-center pb-5 border-b-2 border-gray-line mt-8 lg:mt-14 font-medium">
      <p class="text-lg"> Desafios Completos </p>
      <p class="text-2xl">0</p>
    </div>
    <div class="flex justify-center items-center mt-8 lg:mt-14 text-9xl text-title font-rajdhani">
      <p class="w-1/2 flex gap-x-1"></p><span class="bg-background px-2">:</span>
      <p class="w-1/2 flex gap-x-1"></p>
    </div><button disabled="" class="button completed"> Ciclo completo </button>
  </div>
  <div class="bg-white rounded-md flex flex-1 w-full mt-8 lg:mt-0 shadow-sm w-full lg:w-1/2" id="challenge">
    <section class="flex flex-col flex-1 justify-center items-center p-10 md:p-16">
      <h1 class="font-medium text-center text-2xl"> Finalize um ciclo para receber um desafio </h1><img src="/icons/level-up.svg" alt="Level up" class="mt-10 md:mt-16 min-h-[70px]">
      <p class="text-center mt-4"> Avance de Level completando desafios </p>
    </section>
    <transition-stub data-v-68482f27="" name="slide-fade" appear="false" persisted="false" css="true">
      <!--v-if-->
    </transition-stub>
  </div>
</section>"
`;

exports[`page index > Snapshots > should render all child components and start a cycle button 1`] = `
"<section class="flex flex-col lg:flex-row flex-1 lg:flex-none lg:mt-16 sm:gap-x-10 md:gap-x-20">
  <div class="flex flex-col w-full lg:w-1/2">
    <div class="flex items-center"><img src="https://github.com/kaicLimaOliveira.png" alt="Kaic Dev" class="mr-6 w-20 h-20 rounded-full">
      <div class="flex flex-col w-full"><strong class="text-title text-2xl">Kaic Dev</strong>
        <p class="flex items-center"><img src="/icons/level.svg" alt="level" class="mr-2"><span class="text-base">Level 1</span></p>
      </div>
    </div>
    <div class="flex justify-between items-center pb-5 border-b-2 border-gray-line mt-8 lg:mt-14 font-medium">
      <p class="text-lg"> Desafios Completos </p>
      <p class="text-2xl">0</p>
    </div>
    <div class="flex justify-center items-center mt-8 lg:mt-14 text-9xl text-title font-rajdhani">
      <p class="w-1/2 flex gap-x-1"></p><span class="bg-background px-2">:</span>
      <p class="w-1/2 flex gap-x-1"></p>
    </div><button class="button start"> Iniciar ciclo </button>
  </div>
  <div class="bg-white rounded-md flex flex-1 w-full mt-8 lg:mt-0 shadow-sm w-full lg:w-1/2" id="challenge">
    <section class="flex flex-col flex-1 justify-center items-center p-10 md:p-16">
      <h1 class="font-medium text-center text-2xl"> Finalize um ciclo para receber um desafio </h1><img src="/icons/level-up.svg" alt="Level up" class="mt-10 md:mt-16 min-h-[70px]">
      <p class="text-center mt-4"> Avance de Level completando desafios </p>
    </section>
    <transition-stub data-v-68482f27="" name="slide-fade" appear="false" persisted="false" css="true">
      <!--v-if-->
    </transition-stub>
  </div>
</section>"
`;
